feat: add label selector to the Cluster scale subresource

sebv004 opened the pull request:
# Add VPA Support for CloudNativePG Clusters

## Summary

This PR enables Vertical Pod Autoscaler (VPA) to discover and monitor CloudNativePG database pods by exposing the pod label selector through the Cluster CRD's scale subresource.

## Problem Statement

Currently, VPA cannot provide resource recommendations for CNPG Cluster resources. When attempting to create a VPA targeting a CNPG Cluster, the VPA fails with the following errors:

```yaml
status:
  conditions:
  - type: ConfigUnsupported
    status: "True"
    message: "Cannot read targetRef. Reason: Unhandled targetRef postgresql.cnpg.io/v1 / Cluster / <name>, 
              last error Resource has an empty selector for scale sub-resource"
  - type: NoPodsMatched
    status: "True"
  - type: RecommendationProvided
    status: "False"
```

The root cause is that the Cluster CRD's scale subresource does not expose a pod selector, which VPA requires to discover which pods are managed by the resource.

## Solution

This PR implements VPA support by:

1. **Adding `SelectorLabels` field** to `ClusterStatus` to store the pod label selector
2. **Populating the selector** in the cluster controller with the labels that CNPG already applies to pods:
   ```
   cnpg.io/cluster=<cluster-name>,cnpg.io/podRole=instance
   ```
3. **Exposing the selector** through the scale subresource by adding `labelSelectorPath: .status.selectorLabels`

### Changes Made

#### 1. API Changes (`api/v1/cluster_types.go`)
- Added `SelectorLabels string` field to `ClusterStatus` struct
- Updated kubebuilder marker to include `selectorpath=.status.selectorLabels` in the scale subresource configuration

#### 2. Controller Changes (`internal/controller/cluster_status.go`)
- Modified `updateResourceStatus()` to populate `cluster.Status.SelectorLabels` during reconciliation
- Selector format: `"cnpg.io/cluster=<name>,cnpg.io/podRole=instance"`

#### 3. CRD Changes (`config/crd/bases/postgresql.cnpg.io_clusters.yaml`)
- Added `labelSelectorPath: .status.selectorLabels` to the scale subresource
- Added `selectorLabels` field definition to the status schema

## Testing

Comprehensive testing was performed in a KIND cluster with VPA installed:

### Before Fix (Unpatched CNPG 1.27.1)

Scale subresource response:
```json
{
  "status": {
    "replicas": 1
    // No "selector" field
  }
}
```

VPA Status:
- ❌ `ConfigUnsupported: True`
- ❌ `NoPodsMatched: True`
- ❌ `RecommendationProvided: False`
- ❌ No recommendations

### After Fix (Patched CNPG)

Scale subresource response:
```json
{
  "status": {
    "replicas": 1,
    "selector": "cnpg.io/cluster=test-database,cnpg.io/podRole=instance"
  }
}
```

VPA Status:
- ✅ `RecommendationProvided: True`
- ✅ No errors
- ✅ Actual CPU and memory recommendations provided

VPA Recommender logs confirm successful pod discovery:
```
I1029 14:49:55.115536 cluster_feeder.go:432] "Using selector" 
  selector="cnpg.io/cluster=test-database,cnpg.io/podRole=instance" 
  vpa="test-db/test-database-vpa"
I1029 14:49:55.124309 checkpoint_writer.go:114] "Saved checkpoint for VPA" 
  vpa="test-db/test-database-vpa" container="postgres"
```

## Backward Compatibility

This change is fully backward compatible:
- Only adds new fields, does not modify existing behavior
- The selector labels match the existing labels CNPG already applies to database pods
- Clusters created before this change will have their `selectorLabels` populated on next reconciliation
- No migration or manual intervention required

## Use Case

Users can now create VPAs for CNPG database clusters to get automatic resource recommendations:

```yaml
apiVersion: autoscaling.k8s.io/v1
kind: VerticalPodAutoscaler
metadata:
  name: my-database-vpa
  namespace: default
spec:
  targetRef:
    apiVersion: postgresql.cnpg.io/v1
    kind: Cluster
    name: my-database
  updatePolicy:
    updateMode: "Off"
```

VPA will now successfully discover the database pods and provide recommendations for CPU and memory sizing.

## Benefits

1. **Automatic Resource Optimization**: Users can leverage VPA to right-size their database resources
2. **Cost Efficiency**: Better resource allocation based on actual usage patterns
3. **Operational Excellence**: Reduces manual effort in capacity planning
4. **Standard Kubernetes Integration**: Works with existing VPA deployments

## Related Issues

This addresses the long-standing limitation where VPA could not be used with CNPG Cluster resources, requiring users to either:
- Manually monitor and adjust resources
- Create workarounds with custom label selectors
- Use VPA with StatefulSets directly (bypassing the Cluster abstraction)

## Checklist

- [x] Changes are backward compatible
- [x] API changes include proper documentation comments
- [x] CRD manifests regenerated with `make manifests`
- [x] Deepcopy code regenerated with `make generate`
- [x] Tested in a real Kubernetes cluster with VPA
- [x] No changes to existing functionality
- [x] Labels used match existing CNPG pod labels

## Additional Notes

The implementation uses the existing labels that CNPG applies to all database pods:
- `cnpg.io/cluster`: identifies which cluster the pod belongs to
- `cnpg.io/podRole`: set to "instance" for database pods

This ensures consistency with CNPG's existing labelling scheme and makes the selector reliable for pod discovery.

# Commit message

The `Cluster` scale subresource exposed `spec.instances` and `status.instances` but no selector. Standard Kubernetes tooling expects the scale subresource to publish a label selector identifying the pods it counts; without one, autoscalers cannot map a `Cluster` to its instance pods.

Populate a `status.selector` field and wire it into the scale subresource:

- Add the `status.selector` field, set it from `updateResourceStatus` via the new `Cluster.GetInstancesSelector()`, which returns the canonical serialised selector (`cluster=<name>,role=instance`) built with `labels.SelectorFromSet`.
- Add `selectorpath=.status.selector` to the `+kubebuilder:subresource:scale` marker and regenerate the CRD and API docs.

This makes a `Cluster` a valid `targetRef` for the Vertical Pod Autoscaler (`VPA`), which can now observe the instance pods and emit CPU/memory recommendations, and lets a `HorizontalPodAutoscaler` (`HPA`) scale `spec.instances`.

Document recommended `VPA` usage (recommendation-only `updateMode: Off`, with a warning against `Auto`/`Recreate`/`Initial` to avoid ping-pong restarts against the operator's drift reconciliation) and `HPA` caveats for a stateful database. Add a unit test that verifies the selector format and that it matches the labels applied to every instance pod.

Closes #2574

sebv004 committed:
fix: Enable VPA support by exposing label selector in scale subresource

This fixes compatibility with Vertical Pod Autoscaler (VPA) by properly
implementing the scale subresource selector field, which is required for
VPA to discover and monitor CloudNativePG database pods.

Problem:
VPA was unable to provide resource recommendations for CNPG Cluster resources
because the scale subresource did not expose a pod selector. VPA requires this
selector to discover which pods are managed by a resource, resulting in:
- ConfigUnsupported: True
- NoPodsMatched: True
- RecommendationProvided: False

Root Cause:
The scale subresource was implemented but incomplete - it was missing the
labelSelectorPath configuration required for proper Kubernetes interoperability.

Solution:
1. Added SelectorLabels field to ClusterStatus to store the pod label selector
2. Modified cluster controller to populate this field with pod labels:
   "cnpg.io/cluster=<name>,cnpg.io/podRole=instance"
3. Updated CRD scale subresource configuration to expose the selector via
   labelSelectorPath pointing to .status.selectorLabels

Implementation:
- API: Added SelectorLabels string field to ClusterStatus struct
- Controller: Populates selectorLabels during cluster status reconciliation
- CRD: Added labelSelectorPath to scale subresource configuration

Testing:
Validated with VPA in a KIND cluster. After the fix:
- Scale subresource correctly exposes:
  status.selector="cnpg.io/cluster=<name>,cnpg.io/podRole=instance"
- VPA successfully discovers database pods
- VPA provides resource recommendations (CPU/memory)
- RecommendationProvided: True with actual recommendations

This change is backward compatible as it only adds new fields and does not
modify existing behavior. The selector labels match the existing labels that
CNPG already applies to database pods.

Fixes: VPA compatibility with CNPG Cluster resources
Signed-off-by: Sebastien Villanueva <sebastien.villanueva@gaming1.com>
Signed-off-by: Sebastien Villanueva <sebastien.villanueva@gmail.com>

sdesbure commented:
I don't know how to help but for me it would be interesting (mainly to have "recommended" values for requests and limits)

armru committed:
test: cover selector format for scale subresource

Add a contract assertion that the scale-subresource selector format
(cnpg.io/cluster=<name>,cnpg.io/podRole=instance) parses as a valid
Kubernetes label selector and matches the labels that the operator
applies to instance pods. This locks in the contract that VPA/HPA
rely on for pod discovery.

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru committed:
refactor: align scale-subresource selector with Kubernetes conventions

Build on the VPA support work with naming and idiom cleanups:

- Rename ClusterStatus.SelectorLabels to Selector. Kubernetes scalable
  types (ScaleStatus.Selector) use "selector" for the serialized label
  selector; matching the convention now avoids a deprecation cycle
  later.
- Build the serialized selector with labels.SelectorFromSet instead of
  hand-rolled fmt.Sprintf. The helper sorts requirements by key, so
  the output stays deterministic.
- Regenerate the Cluster CRD and the API reference documentation and
  refresh the spellcheck wordlist accordingly.

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru committed:
docs: document VPA and HPA usage for Cluster resources

Add a Vertical Pod Autoscaler section to the resource-management page
explaining that a Cluster is now a valid VPA targetRef, recommending
the recommendation-only updateMode, and warning that actively mutating
modes (Auto, Recreate, Initial) fight CloudNativePG's pod-spec drift
detection. Include a brief HPA caveat for completeness.

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru commented:
/test

armru committed:
refactor: test the scale selector through the production code path

The unit test for the scale-subresource selector re-derived the expected
selector string with fmt.Sprintf and only checked it against pod labels.
It never invoked the operator's selector-building code, so deleting or
breaking that code would have left the test green.

Extract the selector construction into Cluster.GetInstancesSelector()
(mirroring the adjacent GetServiceReadName/GetServiceReadWriteName helpers
that updateResourceStatus already uses) and assert the test against that
production method. A mutation that changes the selected labels now fails
the test, which previously passed.

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

gbartolini committed:
docs: minor change

Signed-off-by: Gabriele Bartolini <gabriele.bartolini@enterprisedb.com>

gbartolini commented:
/test

gbartolini committed:
docs: minor changes

Signed-off-by: Gabriele Bartolini <gabriele.bartolini@enterprisedb.com>

gbartolini commented:
Tested locally using the usual `cluster-example` and the recommend vpa resource. Neat!

```
$    kubectl describe vpa cluster-example-vpa
Name:         cluster-example-vpa
Namespace:    default
Labels:       <none>
Annotations:  <none>
API Version:  autoscaling.k8s.io/v1
Kind:         VerticalPodAutoscaler
Metadata:
  Creation Timestamp:  2026-06-04T06:07:00Z
  Generation:          1
  Resource Version:    4038
  UID:                 843c7b9a-5595-4643-a24b-5fcc165e78c0
Spec:
  Target Ref:
    API Version:  postgresql.cnpg.io/v1
    Kind:         Cluster
    Name:         cluster-example
  Update Policy:
    Update Mode:  Off
Status:
  Conditions:
    Last Transition Time:  2026-06-04T06:14:07Z
    Status:                True
    Type:                  RecommendationProvided
  Observed Generation:     1
  Recommendation:
    Container Recommendations:
      Container Name:  postgres
      Lower Bound:
        Cpu:     25m
        Memory:  250Mi
      Target:
        Cpu:     25m
        Memory:  250Mi
      Uncapped Target:
        Cpu:     25m
        Memory:  250Mi
      Upper Bound:
        Cpu:     25m
        Memory:  250Mi
Events:          <none>
```

gbartolini committed:
docs: add VPA/HPA to operator capability levels page

Signed-off-by: Gabriele Bartolini <gabriele.bartolini@enterprisedb.com>

gbartolini commented:
Apologies, I had wrongly assigned the pooler's counterpart issue to this PR. Fixed it now with #2574

gbartolini commented:
This can be backported too IMO.

mnencia committed:
docs: correct VPA/HPA guidance for the scale subresource

The committed VPA warning described active update modes as causing
ping-pong restarts, with the operator detecting drift from a VPA-mutated
pod and re-creating it. CloudNativePG does not behave that way: drift
detection compares the stored pod-spec annotation against a target
rebuilt from spec.resources and never reads a pod's live resources, so
an active VPA mode silently diverges the running pod from spec.resources
instead of triggering a rollout.

Reword the warning around that divergence: a shared_buffers value sized
against the declared memory request stops matching the pod's real
limits, and the default primary PodDisruptionBudget blocks eviction of
the primary so VPA stalls. Also scope the HPA webhook-floor caveat to
synchronous replicas, fix the VPA documentation link, and point
per-container recommendations at the postgres container.

Assisted-by: Claude
Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
docs: silence woke master rule on the VPA repo link

The kubernetes/autoscaler repository still uses master as its default
branch, so the VPA documentation link tripped the woke check. Add the
same wokeignore comment the cluster-autoscaler link in scheduling.md
already uses.

Assisted-by: Claude
Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

slocke716 commented:
This is excellent! Thank you for implementing! Are there any gotchas that I should be concerned about using vpa on my clusters? I know this will not automatically update any postgres settings for instance but it seems like just increasing memory and cpu should suffice for most traffic surges
